x-pack/plugins/inference/server/chat_complete/adapters/gemini/gemini_adapter.ts

```diff
@@ -0,0 +1,234 @@
+    };
+  } else if (toolChoice) {
+    return {
+      mode: 'ANY',
+      allowedFunctionNames: [toolChoice.function],
+    };
+  }
+  return undefined;
+}
+
+function toolsToGemini(tools: ToolOptions['tools']): Gemini.Tool[] {
+  return tools
+    ? [
+        {
+          functionDeclarations: Object.entries(tools ?? {}).map(
+            ([toolName, { description, schema }]) => {
+              return {
+                name: toolName,
+                description,
+                parameters: schema
+                  ? toolSchemaToGemini({ schema })
+                  : {
+                      type: Gemini.FunctionDeclarationSchemaType.OBJECT,
+                      properties: {},
+                    },
+              };
+            }
+          ),
+        },
+      ]
+    : [];
+}
+
+function toolSchemaToGemini({ schema }: { schema: ToolSchema }): Gemini.FunctionDeclarationSchema {
+  const convertSchemaType = ({
+    def,
+  }: {
+    def: ToolSchemaType;
+  }): Gemini.FunctionDeclarationSchemaProperty => {
+    switch (def.type) {
+      case 'array':
+        return {
+          type: Gemini.FunctionDeclarationSchemaType.ARRAY,
+          description: def.description,
+          items: convertSchemaType({ def: def.items }) as Gemini.FunctionDeclarationSchema,
+        };
+      case 'object':
+        return {
+          type: Gemini.FunctionDeclarationSchemaType.OBJECT,
+          description: def.description,
+          required: def.required as string[],
+          properties: Object.entries(def.properties).reduce<
+            Record<string, Gemini.FunctionDeclarationSchema>
+          >((properties, [key, prop]) => {
+            properties[key] = convertSchemaType({ def: prop }) as Gemini.FunctionDeclarationSchema;
+            return properties;
+          }, {}),
+        };
+      case 'string':
+        return {
+          type: Gemini.FunctionDeclarationSchemaType.STRING,
+          description: def.description,
+          enum: def.enum ? (def.enum as string[]) : def.const ? [def.const] : undefined,
+        };
+      case 'boolean':
+        return {
+          type: Gemini.FunctionDeclarationSchemaType.BOOLEAN,
+          description: def.description,
+          enum: def.enum ? (def.enum as string[]) : def.const ? [def.const] : undefined,
+        };
+      case 'number':
+        return {
+          type: Gemini.FunctionDeclarationSchemaType.NUMBER,
+          description: def.description,
+          enum: def.enum ? (def.enum as string[]) : def.const ? [def.const] : undefined,
+        };
+    }
+  };
+
+  return {
+    type: Gemini.FunctionDeclarationSchemaType.OBJECT,
+    required: schema.required as string[],
+    properties: Object.entries(schema.properties).reduce<
+      Record<string, Gemini.FunctionDeclarationSchemaProperty>
+    >((properties, [key, def]) => {
+      properties[key] = convertSchemaType({ def });
+      return properties;
+    }, {}),
+  };
+}
+
+function messagesToGemini({
```

Comment:
IIRC, Gemini requires multi-turn messages (user, assistant, user, assistant, ad infinitum). OpenAI doesn't have this guarantee (I think). Should we guarantee multi-turn here, not by throwing, but injecting a user or assistant message? Or maybe we just throw top-level if not multi-turn so we can throw consistently?

Reply:
I wasn't aware of that, but now it explains why some of my tests workflows from yesterday were failing, as I was chaining user messages. 

Yeah, we should probably do something about it. For the API's consistency's sake, I think I would rather inject messages to guarantee consistent rather than throwing... But we would also like to warn the developer that was they are doing might not be optimal for the model their are using... so ideally logging a warning or something on top, maybe?

WDYT?

Reply:
Hum actually, what is done in the connector is that messages from the same role are merged

https://github.com/elastic/kibana/blob/0be8295ed0028870074772c8d82093376482c4eb/x-pack/plugins/stack_connectors/server/connector_types/gemini/gemini.ts#L364-L383

(but only for the `user` role and if not using the `parts` "raw" format).

I guess I can simply do the same in the adapter, by regrouping consecutive same-actor parts under a single message.

Reply:
Done in https://github.com/elastic/kibana/pull/191292/commits/fe00ac639cc14b28c4c22b9c5f47db82e306dc12, and tested it, multiple `text` parts, or mixed `text` + `functionResponse` parts are correctly interpreted by the model